### DISCUSSION TITLE
Running on mobile

### DISCUSSION THREAD

**Discussion by Shoshin23** [openai/whisper#11] [Q&A] [answered]
Hi, 

Awesome work with Whisper and many thanks for putting this out there. It definitely will advance a lot of speech-related applications. I'm building a smartglasses tool that helps people with a visual impairment and one of the most preferred forms of interaction for blind users is through voice. I would like to explore Whisper to run on our Android-based Glasses so want to check If there's an Android port in the works? If not, any high-level ideas on how we can adopt this model on run on mobile? 

Thanks!

**Comment by jongwook** [comment, marked as answer]
We haven't tried mobile applications, but you could try [TorchScript](https://pytorch.org/docs/stable/jit.html) + [PyTorch mobile](https://pytorch.org/mobile/home/) or [ONNX export](https://pytorch.org/docs/stable/onnx.html) + [ONNX mobile runtime](https://onnxruntime.ai/docs/tutorials/mobile/). I expect some difficulties porting the decoding logic, if the mobile environment does not allow running Python.

**Reply by szeidner** [reply]
I am very interested in a mobile application of this as well - for both iOS and Android. I'm currently running Vosk in a mobile app which works quite well, but does not have the same accuracy as Whisper. @Shoshin23 would you mind sharing if you make any progress in this area? I may give it a try as well.

**Reply by edgartaor** [reply]
On the other hand, hardware requirements are high. On mobile, I think, you can only use small or medium models. Is that correct?

**Reply by Redsandro** [reply]
If it helps, python-torch 2.0.1 is on [Termux](https://termux.dev/en/):

https://github.com/termux/termux-packages/tree/master/packages/python-torch

**Comment by nyadla-sys** [comment]
Build something on top of this to run on Android
https://colab.research.google.com/github/usefulsensors/openai-whisper/blob/main/notebooks/tinynn_pytorch_to_tflite_int8.ipynb

**Reply by crazy4pi314** [reply]
Oh this is neat! we were looking for ways to get this working on mobile for an OSS app for local transcription for journalists. Do you know if tflite can leverage the special hardware acceleration on apple/google phones?

**Comment by nyadla-sys** [comment]
Refer below for tflite model files
https://github.com/usefulsensors/openai-whisper/tree/main/models

**Comment by nyadla-sys** [comment]
@zhuzilin @ArtyomZemlyak
I could generate encoder int8 tflite model from openai->whisper(pytorch) and ran inference.
https://colab.research.google.com/github/usefulsensors/openai-whisper/blob/main/notebooks/tinynn_pytorch_to_tflite_int8.ipynb
can someone help me to generate int8 decoder tflite model from openai->whisper(pytorch)?

**Comment by leecrossley** [comment]
I got Whisper working on iOS (android is probably easier) by converting the (small) model to CoreML packages in python with the coremltools convert function, as well as writing quite a bit of Swift to them in my scenario. It's not super fast, and the app size is around 500MB using the small model; I haven't looked at the issues on older iOS devices yet (crashing).

A few considerations depend on what you're trying to achieve - there's the speech framework on iOS, for instance (but it recommends an internet connection) which may be faster, more real time, support older devices, reduce app size....

You might want to consider https://github.com/snipsco/snips-platform-android-demo as an option reading your case - just throwing it out there for evaluation

Here's a fun little post about my adventure so far but happy to have a chat to see if I can help: https://sandigital.uk/blog/openai-whisper-scottish-accent/

**Reply by jrappeneker** [reply]
Hi, would you be willing to share your code? I'm working on an app to perform linguistics research and it would be really helpful!

**Reply by Alirsgp** [reply]
I'd also love a way to see this work. Have some use cases in mind.

**Reply by leecrossley** [reply]
I'll get this shared as soon as I get a chance, hopefully this week

**Reply by Alirsgp** [reply]
Thanks lee. I'm a bit stuck on the CoreML conversion process.

**Reply by Shabbir7999** [reply]
@leecrossley  Can you please share your code for Android or IOS with me?

**Comment by nyadla-sys** [comment]
encoder/decoder tflite(int8) models can be generated with the below script
https://colab.research.google.com/github/usefulsensors/openai-whisper/blob/main/notebooks/tinynn_pytorch_to_tflite_int8.ipynb

**Comment by nyadla-sys** [comment]
here is the colab link to generate whisper.tflite file
https://colab.research.google.com/github/usefulsensors/openai-whisper/blob/main/notebooks/tflite_from_huggingface_whisper.ipynb

Updated input shapes using input signature to create a new hybrid whisper.tflite 40MB model (activations are in float32, weights are in int8). I was then able to execute inference with dummy input on whisper.tflite using minimal example from TFLite C++ API on a Linux PC.
Follow the below instructions to build minimal binary on Linux
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/examples/minimal

run the below command to run inference on [whisper.tflite](https://github.com/usefulsensors/openai-whisper/blob/main/models/whisper.tflite)
./minimal ~/Desktop/whisper.tflite

Next steps:Feed the audio input->generate spectrograms->feed to model->post process model output in c++ code

**Comment by nyadla-sys** [comment]
Please follow the README to run whisper.tflite using tflite framework
https://github.com/usefulsensors/openai-whisper

**Comment by Topping1** [comment]
> Hi,
> 
> Awesome work with Whisper and many thanks for putting this out there. It definitely will advance a lot of speech-related applications. I'm building a smartglasses tool that helps people with a visual impairment and one of the most preferred forms of interaction for blind users is through voice. I would like to explore Whisper to run on our Android-based Glasses so want to check If there's an Android port in the works? If not, any high-level ideas on how we can adopt this model on run on mobile?
> 
> Thanks!

In this repository you can find a Flutter port of Whisper: https://github.com/azkadev/whisper_dart
it is based on the C++ port of whisper: https://github.com/ggerganov/whisper.cpp

**Comment by nyadla-sys** [comment]
I developed Android APP based on whisper.tflite (quantized ~40MB) running inference in ~2 seconds for 30 seconds audio clip
https://github.com/usefulsensors/openai-whisper/blob/main/android_app/release/app-release.apk
Use the below command to push on to android phone 
**adb install -r -t ~/android_app/release/app-release.apk**

**Reply by View-my-Git-Lab-krafi** [reply]
> I developed Android APP based on whisper.tflite (quantized ~40MB) running inference in ~2 seconds for 30 seconds audio clip https://github.com/usefulsensors/openai-whisper/blob/main/android_app/release/app-release.apk Use the below command to push on to android phone **adb install -r -t ~/android_app/release/app-release.apk**

Nice but looks like it's only for English

**Comment by nyadla-sys** [comment]
I released Whisper Android App based on Whisper.tflite ~40MB quantized model to the Android App store. 

Link to Android app store 
https://play.google.com/store/apps/details?id=com.whisper.android.tflitecpp

**Comment by nyadla-sys** [comment]
This is only a proof-of-concept project to create an Android app based on Whisper TFLite, which leverages the stock Android UI to show off its features.
[Whisper-TFLIte-Android-Example](https://github.com/usefulsensors/openai-whisper/tree/main/android_app/Whisper-TFLIte-Android-Example)

**Comment by nyadla-sys** [comment]
https://apps.apple.com/in/app/whisper-asr/id6444556326

Pls download iOS app from apple app store which uses whisper tflite model

**Reply by amgaa** [reply]
Hey, awesome work there! 

I was just wondering if you plan to extend your iOS app to other languages. It would be great to have the app available in more languages for a wider audience. Also, have you considered opening up the source code for the app for others to build their own models? Just curious about your thoughts on this. Thanks!

**Reply by cvladan** [reply]
Yes, would it be possible for us to test a multilingual model to see how it performs? Thank you for considering this request.

**Reply by cvladan** [reply]
On Android

**Comment by nyadla-sys** [comment]
Please feel free to download the openai/whisper-tiny tflite-based Android APP from Google App Store.
https://play.google.com/store/apps/details?id=com.whisper.android.tflitecpp

**Reply by Squishypone** [reply]
Very interesting would it be possible to put this in some sort of open keyboard app so you can use it as a voice to text replacement for the Google voice to text?

**Reply by nyadla-sys** [reply]
Please refer the below link which uses tflite as part of AndroidVoice keyboard
https://github.com/MichaelMcCulloch/WhisperVoiceKeyboard

**Reply by MichaelMcCulloch** [reply]
@nyadla-sys I am attempting to move to the large model, and run inference on NNAPI, and I am very much stuck. I don't suppose you would be able to lend your brain and eyes?

**Reply by nyadla-sys** [reply]
@MichaelMcCulloch are you able to generate large whisper tflite model?if so what kind of issues you are seeing.
Please give me more details along with adb logcat I can help on this

**Reply by MichaelMcCulloch** [reply]
Let's move this conversation to an [issue](https://github.com/MichaelMcCulloch/WhisperVoiceKeyboard/issues/12) on my repo.

**Reply by slurpey** [reply]
I would be definitely willing to pay for your app if it had an automated copy of the text to the clipboard or a big copy button at the bottom, so I could use it as a replacement for Google Voice Typing.

**Comment by SubhamSubhasisPatra** [comment]
Hey @nyadla-sys, I just came across your application it's awesome, and I want to integrate it into a react native application. Can u plz give me some clues on how to integrate it?

**Comment by nyadla-sys** [comment]
You may find the source code for a Linux C++ application and an Android Java app on my Github repository at the following link: https://github.com/usefulsensors/openai-whisper.

**Reply by SubhamSubhasisPatra** [reply]
@nyadla-sys, This .tflite will be easy to use directly in an Android application, but is there any way to create the models.json and weight.bin file from that, because react native does not support .tflite?

**Comment by zhanghui19880625** [comment]
hello, Why does the whisper_android app run above android 9.0?
I want to one can run above android 7.0

**Reply by nyadla-sys** [reply]
Android example to use whisper tflite model https://github.com/nyadla-sys/whisper.tflite/tree/main

**Comment by tawanda-profuse** [comment]
Which app are you referring to specifically? Please share the link.

**Reply by slurpey** [reply]
It's the one above my reply trail: https://play.google.com/store/apps/details?id=com.whisper.android.tflitecpp

**Reply by adamlogan** [reply]
Link is dead.

**Comment by View-my-Git-Lab-krafi** [comment]
i dont know do anyown want to use whisper using web browser . and your pc as server https://gitlab.com/krafi/whisperweb
